pkg/resource/firewall/hook.go

```diff
@@ -26,4 +35,259 @@ var (
+}
+
+// addPutFieldsToSpec will describe logging config and add its
+// returned values to the Firewall spec.
+func (rm *resourceManager) addLoggingConfigToSpec(
+	ctx context.Context,
+	r *resource,
+	ko *svcapitypes.Firewall,
+) (err error) {
+	getLoggingConfigurationResponse, err := rm.sdkapi.DescribeLoggingConfigurationWithContext(ctx, rm.getLoggingConfigurationPayload(r))
+	if err != nil {
+		return err
+	}
+	ko.Spec.LoggingConfiguration = rm.setResourceLoggingConfiguration(r, getLoggingConfigurationResponse)
+	return nil
+}
+
+func customPreCompare(
+	a *resource,
+	b *resource,
+) {
+	if a.ko.Spec.LoggingConfiguration == nil && b.ko.Spec.LoggingConfiguration == nil {
+		return
+	}
+
+	if a.ko.Spec.LoggingConfiguration == nil {
+		a.ko.Spec.LoggingConfiguration = &svcapitypes.LoggingConfiguration{}
+	}
+
+	if b.ko.Spec.LoggingConfiguration == nil {
+		b.ko.Spec.LoggingConfiguration = &svcapitypes.LoggingConfiguration{}
+	}
+
+	sort.Slice(a.ko.Spec.LoggingConfiguration.LogDestinationConfigs[:], func(i, j int) bool {
+		return *a.ko.Spec.LoggingConfiguration.LogDestinationConfigs[i].LogType < *a.ko.Spec.LoggingConfiguration.LogDestinationConfigs[j].LogType
+	})
+	sort.Slice(b.ko.Spec.LoggingConfiguration.LogDestinationConfigs[:], func(i, j int) bool {
+		return *b.ko.Spec.LoggingConfiguration.LogDestinationConfigs[i].LogType < *b.ko.Spec.LoggingConfiguration.LogDestinationConfigs[j].LogType
+	})
+}
+
+func (rm *resourceManager) getLoggingConfigurationPayload(
+	r *resource,
+) *svcsdk.DescribeLoggingConfigurationInput {
+	res := &svcsdk.DescribeLoggingConfigurationInput{}
+	res.SetFirewallName(*r.ko.Spec.FirewallName)
+	return res
+}
+
+func (rm *resourceManager) newLoggingConfigurationPayload(
+	r *resource,
+) *svcsdk.UpdateLoggingConfigurationInput {
+	res := &svcsdk.UpdateLoggingConfigurationInput{}
+	res.SetFirewallName(*r.ko.Spec.FirewallName)
+	if r.ko.Spec.LoggingConfiguration != nil {
+		res.SetLoggingConfiguration(rm.newLoggingConfiguration(r))
+	} else {
+		res.SetLoggingConfiguration(&svcsdk.LoggingConfiguration{})
+	}
+	return res
+}
+
+func makeLogDestinationConfigMapKey(cfg *svcapitypes.LogDestinationConfig) string {
+	keys := make([]string, 0, len(cfg.LogDestination))
+	for k := range cfg.LogDestination {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	key := "LogDestination:"
+	for _, k := range keys {
+		v := *cfg.LogDestination[k]
+		key += fmt.Sprintf("%s:%s,", k, v)
+	}
+
+	if cfg.LogDestinationType != nil {
+		key += fmt.Sprintf("LogDestinationType:%s,", *cfg.LogDestinationType)
+	}
+	if cfg.LogType != nil {
+		key += fmt.Sprintf("LogType:%s", *cfg.LogType)
+	}
+	return key
+}
+
+func makeLogDestinationConfigMap(logDestinationConfig []*svcapitypes.LogDestinationConfig) map[string]*svcapitypes.LogDestinationConfig {
+	logDestinationConfigMap := make(map[string]*svcapitypes.LogDestinationConfig)
+
+	for _, cfg := range logDestinationConfig {
+		logDestinationConfigMap[makeLogDestinationConfigMapKey(cfg)] = cfg
+	}
```

Comment:
I am using the map value 'cfg' in compareLoggingDestinationConfigs function, so we can not make value of the map as nil. 

Let me explain the purpose of this map. 

In compareLoggingDestinationConfigs() , we have to compare arrays of type []*svcapitypes.LogDestinationConfig. `LogDestinationConfig` in itself is complex type which has a map (`LogDestination`) as one of its fields. This makes comparison of these 2 arrays difficult.

So, to compare these arrays, I convert the array into a map, because map comparisons are easier to handle based on map keys. So, to convert each array object into a map entry, we need to have a key corresponding to that  array object. `makeLogDestinationConfigMapKey` creates this key from each array object. The value associated with this key is the array object itself, so that it is easier to retrieve the array object.

So, in summary,
1. `makeLogDestinationConfigMapKey` gives unique key for each array entry of type `*svcapitypes.LogDestinationConfig`.
2. `makeLogDestinationConfigMap` creates map with its key provided by `makeLogDestinationConfigMapKey` and its value as the array entry of type `svcapitypes.LogDestinationConfig`
3. In `compareLoggingDestinationConfigs` , we create 2 maps. One that represents what is desired logging configuration and one that represents what is existing logging configuration.
Now that, we have 2 maps, we iterate over `desired` map and check if the keys are present in `latest` map. If they are not present, we append the `value` of that map (which is `*svcapitypes.LogDestinationConfig`) into the list of configs to be `added`. Same case to determine list of configs to be `deleted`.

Reply:
I understand the problem, this is something we've seen with other APIs, but had to handle with some custom comparison/update functions. But in your case looks like it's a bit different.
I'm going to clone your branch and play a little bit with the implementation exploring the complexity of this case.

Reply:
Let me know if you got a chance to check more on this one.